cpp/src/arrow/util/small_vector.h

```diff
@@ -0,0 +1,499 @@
+  }
+
+  StaticVectorImpl(std::vector<T>&& other) noexcept {  // NOLINT: explicit
+    init_by_moving(other.size(), other.data());
+  }
+
+  StaticVectorImpl& operator=(const std::vector<T>& other) {
+    assign_by_copying(other.size(), other.data());
+    return *this;
+  }
+
+  StaticVectorImpl& operator=(std::vector<T>&& other) noexcept {
+    assign_by_moving(other.size(), other.data());
+    return *this;
+  }
+
+  // Constructing from count and optional initialization value
+  explicit StaticVectorImpl(size_t count) {
+    storage_.bump_size(count);
+    auto* p = data_ptr();
+    for (size_t i = 0; i < count; ++i) {
+      new (&p[i]) T();
+    }
+  }
+
+  StaticVectorImpl(size_t count, const T& value) {
+    storage_.bump_size(count);
+    auto* p = data_ptr();
+    for (size_t i = 0; i < count; ++i) {
+      new (&p[i]) T(value);
+    }
+  }
+
+  StaticVectorImpl(std::initializer_list<T> values) {
+    storage_.bump_size(values.size());
+    auto* p = data_ptr();
+    for (auto&& v : values) {
+      // XXX cannot move initializer values?
+      new (p++) T(v);
+    }
+  }
+
+  constexpr bool empty() const { return storage_.size_ == 0; }
+
+  constexpr size_t size() const { return storage_.size_; }
+
+  constexpr size_t capacity() const { return storage_.capacity(); }
+
+  constexpr size_t max_size() const { return storage_.max_size(); }
+
+  T& operator[](size_t i) { return data_ptr()[i]; }
+
+  constexpr const T& operator[](size_t i) const { return const_data_ptr()[i]; }
+
+  T& front() { return data_ptr()[0]; }
+
+  constexpr const T& front() const { return const_data_ptr()[0]; }
+
+  T& back() { return data_ptr()[storage_.size_ - 1]; }
+
+  constexpr const T& back() const { return const_data_ptr()[storage_.size_ - 1]; }
+
+  T* data() { return data_ptr(); }
+
+  constexpr const T* data() const { return const_data_ptr(); }
+
+  iterator begin() { return iterator(data_ptr()); }
+
+  constexpr const_iterator begin() const { return const_iterator(const_data_ptr()); }
+
+  iterator end() { return iterator(data_ptr() + storage_.size_); }
+
+  constexpr const_iterator end() const {
+    return const_iterator(const_data_ptr() + storage_.size_);
+  }
+
+  void reserve(size_t n) { storage_.reserve(n); }
+
+  void clear() { storage_.clear(); }
+
+  void push_back(const T& value) {
+    storage_.bump_size(1);
+    new (data_ptr() + storage_.size_ - 1) T(value);
+  }
+
+  void push_back(T&& value) {
+    storage_.bump_size(1);
+    new (data_ptr() + storage_.size_ - 1) T(std::move(value));
+  }
+
+  template <typename... Args>
+  void emplace_back(Args&&... args) {
+    storage_.bump_size(1);
+    new (data_ptr() + storage_.size_ - 1) T(std::forward<Args>(args)...);
+  }
+
+  template <typename InputIt>
+  iterator insert(const_iterator insert_at, InputIt first, InputIt last) {
+    const size_t n = storage_.size_;
+    const size_t it_size = static_cast<size_t>(last - first);  // XXX might be O(n)?
```

Comment:
No, actually only RandomAccessIterators support `operator-` (so this would fail to compile for std::list::iterator).

For other types you'd have to use `std::distance` which can be O(n). Most vector implementations have different code paths for different `typename std::iterator_traits<Iter>::iterator_category`, only some of which presize.

Altogether I think it's preferable to use `std::distance` here an forgo support for input iterators

Reply:
> Altogether I think it's preferable to use std::distance here an forgo support for input iterators

Hmm... if we want to forgo support for non-random iterators, we should keep using substraction, no?

Reply:
sure, that's also acceptable

Reply:
(IIUC the most common non-random access iterators we use in the arrow codebase are `unordered_{set,map}::iterator`, so the consequence here would be `small_vec.insert(small_vec.begin() + 3, set.begin(), set.end())` won't compile)

Reply:
What would you prefer?

Reply:
I think it's fine to leave support for non-random access iterators for follow up. We can add that when we need it